pool/entities.go

```diff
@@ -0,0 +1,170 @@
+package pool
+
+type ID string
+
+type Name string
+
+// I don't know i have to bring Peer Available storage and other metric inside Peer
+type Peer string
+
+type ReplicationFactor int8
+
+// Requirement of joining the pool
+type Requirement struct {
+	Latency   string `json:"latency"`
+	Bandwidth string `json:"bandwidth"`
+	ReplicationFactor ReplicationFactor `json:"rep_factor"`
+}
+
+// ResourceRequest object (it is much more complicated for simplicity we think as file).
+type Resource struct {
+	ID  ID `json:"id"`
+	FileName  Name
+	Owner     Peer
+	Size      string
+	ReplicationFactor ReplicationFactor
+	Allocation []Peer
+}
+
+
+// Event is a domain event marker.
+type Event interface {
+	isEvent()
+}
+
+func (e PoolCreated) isEvent() {}
+func (e ResourceRequested) isEvent() {}
+func (e ResourceAllocated) isEvent() {}
+func (e MemberJoined) isEvent() {}
+func (e MemberLeaved) isEvent() {}
+
+// PoolCreated Event
+type PoolCreated struct {
+	ID	         ID           `json:"id"`
+	Name         Name         `json:"name"`
+	Creator      Peer         `json:"creator"`
+	Members      []Peer       `json:"new_creator"`
+	Requirement  Requirement  `json:"new_requirement"`
+}
+
+// ResourceRequested Event
+type ResourceRequested struct {
+	ID               ID               `json:"id"`
+	ResourceRequest  Resource         `json:"new_resource"`
+}
+
+// ResourceAllocated Event
+type ResourceAllocated struct {
+	ID          ID          `json:"id"`
+	Resource  ID  `json:"new_resource"`
+	Allocation Peer
+}
+
+// MemberJoined Event
+type MemberJoined struct {
+	ID      ID    `json:"id"`
+	Member  Peer  `json:"member"`
+}
+
+// MemberLeaved Event
+type MemberLeaved struct {
+	ID      ID    `json:"id"`
+	Member  Peer  `json:"member"`
```

Comment:
Do we also need a PoolID in the structure? to ensure the message is sent in the correct pool of the user is a member of multiple pools, and for debuggin purposes?

Reply:
Please be more focused on business.

Reply:
What does this mean Farhoud?

Reply:
@ehsan6sha 
## EventSourcing:
Instead of storing the state on the store, we keep a sequence of changes (Events) on the store and calculate the current State from digesting the events.
ps. If you put data in a single event store they call it event sourcing if you keep the data on the linked list and copy it into every node they call it blockchain.

## Aggregate
Root aggregate is the state of application for a specific service. I Imagined our service will manage the Pool so the root aggregate is the Pool.
Every event carries the Pool id with itself.
The impact of every event type on the state can be seen inside the On method.

## References
- https://victoramartinez.com/posts/event-sourcing-in-go/
- https://blog.cleancoder.com/uncle-bob/2012/08/13/the-clean-architecture.html